File: backend/src/main/java/com/example/backend/service/InAppNotificationService.java
package com.example.backend.service;

import com.example.backend.dto.response.NotificationResponse;
import com.example.backend.entity.enums.NotificationType;
import com.example.backend.entity.enums.Role;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service for managing in-app notifications displayed on the website
 */
public interface InAppNotificationService {

    /**
     * Create a notification for specific user
     * @param userId User ID to receive notification
     * @param type Type of notification
     * @param content Notification content
     * @param linkUrl Optional link URL (e.g., to view the related entity)
     * @return Notification ID
     */
    String createNotificationForUser(String userId, NotificationType type, String content, String linkUrl);

    /**
     * Create a notification for all users with specific role(s)
     * @param roles Target roles
     * @param type Type of notification
     * @param content Notification content
     * @param linkUrl Optional link URL
     * @return Notification ID
     */
    String createNotificationForRoles(Role[] roles, NotificationType type, String content, String linkUrl);

    /**
     * Create a broadcast notification for all admins
     * @param type Type of notification
     * @param content Notification content
     * @param linkUrl Optional link URL
     * @return Notification ID
     */
    String createNotificationForAllAdmins(NotificationType type, String content, String linkUrl);

    /**
     * Get all notifications for a user
     * @param keycloakId User's Keycloak ID
     * @param pageable Pagination
     * @return Page of notifications
     */
    Page<NotificationResponse> getNotificationsForUser(String keycloakId, Pageable pageable);

    /**
     * Get unread notifications for a user
     * @param keycloakId User's Keycloak ID
     * @param pageable Pagination
     * @return Page of unread notifications
     */
    Page<NotificationResponse> getUnreadNotificationsForUser(String keycloakId, Pageable pageable);

    /**
     * Get count of unread notifications for a user
     * @param keycloakId User's Keycloak ID
     * @return Count of unread notifications
     */
    long getUnreadCount(String keycloakId);

    /**
     * Mark a notification as read
     * @param keycloakId User's Keycloak ID
     * @param notificationId Notification ID
     */
    void markAsRead(String keycloakId, String notificationId);

    /**
     * Mark all notifications as read for a user
     * @param keycloakId User's Keycloak ID
     * @return Number of notifications marked as read
     */
    int markAllAsRead(String keycloakId);

    /**
     * Delete old notifications (cleanup)
     * @param daysOld Delete notifications older than this many days
     */
    void deleteOldNotifications(int daysOld);
}

File: website/fe_admin/app/service/inAppNotificationService.ts
import axiosInstance from '../config/axios';
import type { ApiResponse } from './types';

export interface InAppNotification {
  notificationId: string;
  content: string;
  type: string;
  typeDisplayName: string;
  linkUrl: string | null;
  isRead: boolean;
  isBroadcast: boolean;
  createdAt: string;
}

export interface NotificationPage {
  content: InAppNotification[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

/**
 * In-App Notification Service
 * Handles notification bar notifications for the website
 */
class InAppNotificationService {
  /**
   * Get all notifications for current user
   */
  async getMyNotifications(page: number = 0, size: number = 20): Promise<NotificationPage> {
    console.log('Calling getMyNotifications API with params:', { page, size });
    const response = await axiosInstance.get<ApiResponse<NotificationPage>>('/notifications', {
      params: { page, size }
    });
    console.log('API Response:', response.data);
    return response.data.data;
  }

  /**
   * Get unread notifications for current user
   */
  async getUnreadNotifications(page: number = 0, size: number = 20): Promise<NotificationPage> {
    const response = await axiosInstance.get<ApiResponse<NotificationPage>>('/notifications/unread', {
      params: { page, size }
    });
    return response.data.data;
  }

  /**
   * Get unread notification count for badge display
   */
  async getUnreadCount(): Promise<number> {
    const response = await axiosInstance.get<ApiResponse<{ unreadCount: number }>>('/notifications/unread-count');
    return response.data.data.unreadCount;
  }

  /**
   * Mark a notification as read
   */
  async markAsRead(notificationId: string): Promise<void> {
    await axiosInstance.patch<ApiResponse<string>>(`/notifications/${notificationId}/read`);
  }

  /**
   * Mark all notifications as read
   */
  async markAllAsRead(): Promise<number> {
    const response = await axiosInstance.post<ApiResponse<{ markedCount: number }>>('/notifications/mark-all-read');
    return response.data.data.markedCount;
  }
}

export default new InAppNotificationService();
